### PR TITLE
mssmt: add new Copy method and InsertMany to optimize slightly 

### PR DESCRIPTION
**Pull request by Roasbeef**
In this commit, we add a new `Copy` method that takes as an arg another `mssmt.Tree` and inserts all the keys+leaves from the target tree into the source tree. We then optimize a bit by adding an `InsertMany` method that will call `insert` N times in the same `TreeStoreUpdateTx`. This doesn't yet implement any fancy optimizations to ensure the root is only updated once as https://github.com/lightninglabs/taproot-assets/pull/1347 does. 

This was spun off a work unit related to https://github.com/lightninglabs/taproot-assets/pull/1464.

### CODE REVIEW

File: mssmt/compacted_tree.go [hunk 1]
```diff
@@ -392,3 +392,166 @@ func (t *CompactedTree) MerkleProof(ctx context.Context, key [hashSize]byte) (
 
 	return NewProof(proof), nil
 }
+
+// collectLeaves recursively traverses the compacted tree represented by the
+// given node and collects all non-empty leaf nodes along with their keys.
+func collectLeaves(ctx context.Context, tx TreeStoreViewTx, node Node) (
```
Comment (ffranr) [inline mssmt/compacted_tree.go:398]: I don't think this function is used. Is it included intentionally?

File: mssmt/compacted_tree.go [hunk 2]
```diff
@@ -392,3 +392,166 @@ func (t *CompactedTree) MerkleProof(ctx context.Context, key [hashSize]byte) (
...
+	}
+
+	// Check against EmptyTree branches requires depth. If we encounter an
+	// unexpected node that isn't identifiable as empty, return error.
+	return nil, fmt.Errorf("unexpected node type %T encountered "+
+		"during leaf collection", node)
+}
+
+// collectLeavesRecursive is the helper for collectLeaves that includes depth.
+func collectLeavesRecursive(ctx context.Context, tx TreeStoreViewTx, node Node,
```
Comment (ffranr) [inline mssmt/compacted_tree.go:432]: I think this function used to be a helper for `collectLeaves`, but now it is the go-to. So I think doc is stale.

File: mssmt/compacted_tree.go [hunk 3]
```diff
@@ -392,3 +392,166 @@ func (t *CompactedTree) MerkleProof(ctx context.Context, key [hashSize]byte) (
...
+		return make(map[[hashSize]byte]*LeafNode), nil
+	}
+
+	return nil, fmt.Errorf("unexpected node type %T encountered "+
+		"during leaf collection at depth %d", node, depth)
+}
+
+// Copy copies all the key-value pairs from the source tree into the target
+// tree. The target tree is assumed to be empty.
+func (t *CompactedTree) Copy(ctx context.Context, targetTree Tree) error {
```
Comment (ffranr) [inline mssmt/compacted_tree.go:519]: > The target tree is assumed to be empty.

I don't think the target tree has to be empty. So maybe we can remove this from the doc. Otherwise, if it has to be empty, then I think we should assert that in code in addition to having this note in the doc.

File: mssmt/tree.go [hunk 4]
```diff
@@ -333,6 +333,112 @@ func (t *FullTree) MerkleProof(ctx context.Context, key [hashSize]byte) (
...
+		if err != nil {
+			return nil, err
+		}
+
+		// Recursively find leaves in the right subtree. We need to set
+		// the bit corresponding to the current depth in the key prefix.
+		rightKeyPrefix := keyPrefix
+		byteIndex := depth / 8
+		bitIndex := depth % 8
+		rightKeyPrefix[byteIndex] |= (1 << bitIndex)
```
Comment (ffranr) [inline mssmt/tree.go:378]: I think that this should be spun into its own function so that we can add a bit more docs to explain what's happening here. I'm surprised that this function doesn't exist already. It's repeated in `PackBits` for example.

File: mssmt/tree_test.go [hunk 5]
```diff
@@ -822,6 +822,103 @@ func TestBIPTestVectors(t *testing.T) {
 	}
 }
 
+// TestTreeCopy tests the Copy method for both FullTree and CompactedTree,
+// including copying between different tree types.
+func TestTreeCopy(t *testing.T) {
```
Comment (ffranr) [inline mssmt/tree_test.go:827]: All these test cases assume that the target tree is empty. We might want to assert that in the copy routine or add test coverage for non-empty.

**Review by ffranr — COMMENTED** [review]
Looks good to me!

Some of the comments are missing punctuation, but no big deal.

File: mssmt/compacted_tree.go (hunk 1 above)
Comment (Roasbeef) [reply mssmt/compacted_tree.go:398]: Yep not needed, can remove.

File: mssmt/compacted_tree.go (hunk 3 above)
Comment (Roasbeef) [reply mssmt/compacted_tree.go:519]: There's no need for it to be empty. Will update the `godoc` here, and also add additional test cases of non-empty roots.

File: mssmt/tree_test.go (hunk 5 above)
Comment (Roasbeef) [reply mssmt/tree_test.go:829]: Expanded to include a non-empty target tree.

**Review by ffranr — APPROVED** [review]
LGTM :+1:
